### PR TITLE
Fix issue with Sphinx Documentation not automatically updating

### PR DESCRIPTION
**Pull request by ew3361zh**
PR to address issue #863 where Sphinx documentation is not being updated automatically with the current GitHub Actions Workflow.

## There were a few issues going on:
1. The .rst files used to create the html sphinx documentation haven't updated since the first time they were generated. They need to be regularly regenerated to reflect changes in the python files and overall model structure. This PR adds an action to address this but this can be done manually if this seems like too much to do with every push to GitHub. It depends on how up-to-date we want the Sphinx documentation to be.
2. The html documentation was using the old .rst files because new ones hadn't been generated.
3. The changes made by the Sphinx GitHub Action weren't being pushed to the PR after they were run.

## What this PR does to address the above issues:
- Updates the Sphinx Github Action to: 
   - Remove current .rst files except for index.rst and modules.rst.
   - Regenerate Sphinx documentation for the project using the sphinx-apidoc command.
   - Make new html documents based on the updated documentation.
   - Add these Sphinx documentation changes to the PR.

### Notes
- This first PR will be enormous because Sphinx documentation hasn't been updated in almost a year and a lot has changed in the model in the last year. I branched off this branch and the changes were minimal so it's not adding 43k lines every time its running the action.
- This will likely add some documentation to many PRs. We will need to tell people to ignore these files in their reviews.
- The index.rst and modules.rst files caused issues in creating the Sphinx documentation when they were not present during the rebuild (i.e. I tried removing them with all the other .rst files and the sphinx-apidoc command failed because it was looking for these two files).

### CODE REVIEW

File: .github/workflows/sphinx.yml [hunk 1]
```diff
@@ -21,7 +24,18 @@ jobs:
       run: |
         python -m pip install --upgrade pip
         if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
+    - name: Remove Existing .rst Files (excluding index.rst and modules.rst)
+      run: find docs -maxdepth 1 -type f -name '*.rst' ! -name 'index.rst' ! -name 'modules.rst' -exec rm -f {} \;
+    - name: Generate Sphinx Documentation with sphinx-apidoc
+      run: sphinx-apidoc -o docs .
```
Comment (Isgotkowitz) [inline .github/workflows/sphinx.yml]:
```suggestion
      run: sphinx-apidoc -o docs . tests/
```
Adding `tests/` will stop documentation from being generated for all the tests - which I think would be a good thing. Our testing is not very complex, and IMO does not need Sphinx documentation.

**Review by Isgotkowitz — COMMENTED** [review]
This is a massive improvement, nice job Niko. I read through the new html documentation, and it looks like everything is there. There are some ways that I think the organization and verbosity of the documentation could be improved, but I don't know if it would be within the scope of this PR to try and automate those.

Only suggested change is to keep documentation for being generated for the tests, because IMO it is not necessary. However I look forward to hearing others' thoughts on this.

Commenting only for right now because I made a branch off of this one (#1020), pushed it to the remote, and when the Sphinx action failed on that branch it somehow made the Sphinx action on this branch fail too. Once we get whatever is going on with that figured out, I will come back and approve.

**Comment by ew3361zh** [comment]
@Isgotkowitz could you reprint the error message you saw on the Sphinx build from #1020? For some reason I'm not finding it there.

The one currently here is:

Run actions-js/push@master
Started: bash /home/runner/work/_actions/actions-js/push/master/start.sh
Push to branch eds_trial_sphinx_updater
[detached HEAD [13](https://github.com/RuminantFarmSystems/MASM/actions/runs/7103107082/job/19335012054?pr=1010#step:8:14)12e0f0f] Update Sphinx Documentation
 4 files changed, 3 insertions(+), 3 deletions(-)
To https://github.com/RuminantFarmSystems/MASM.git
 ! [rejected]            HEAD -> eds_trial_sphinx_updater (fetch first)
error: failed to push some refs to 'https://github.com/RuminantFarmSystems/MASM.git'
hint: Updates were rejected because the remote contains work that you do not
hint: have locally. This is usually caused by another repository pushing to
hint: the same ref. If you want to integrate the remote changes, use
hint: 'git pull' before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
Error: Invalid status code: 1
    at ChildProcess.<anonymous> (/home/runner/work/_actions/actions-js/push/master/start.js:9:19)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1100:[16](https://github.com/RuminantFarmSystems/MASM/actions/runs/7103107082/job/19335012054?pr=1010#step:8:17))
    at Process.ChildProcess._handle.onexit (node:internal/child_process:304:5) {
  code: 1
}
Error: Invalid status code: 1
    at ChildProcess.<anonymous> (/home/runner/work/_actions/actions-js/push/master/start.js:9:[19](https://github.com/RuminantFarmSystems/MASM/actions/runs/7103107082/job/19335012054?pr=1010#step:8:20))
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1100:16)
    at Process.ChildProcess._handle.onexit (node:internal/child_process:[30](https://github.com/RuminantFarmSystems/MASM/actions/runs/7103107082/job/19335012054?pr=1010#step:8:31)4:5)

File: .github/workflows/sphinx.yml (hunk 1 above)
Comment (ew3361zh) [reply .github/workflows/sphinx.yml]: That makes sense to me. I'd be curious what others think about that as well though. @allisterakun @ltp23 @PooyaHekmati thoughts?

**Comment by Isgotkowitz** [comment]
> @Isgotkowitz could you reprint the error message you saw on the Sphinx build from #1020? For some reason I'm not finding it there.
> 
> The one currently here is:
> 
> Run actions-js/push@master
Started: bash /home/runner/work/_actions/actions-js/push/master/start.sh
Push to branch eds_trial_sphinx_updater
[detached HEAD [13](https://github.com/RuminantFarmSystems/MASM/actions/runs/7103107082/job/19335012054?pr=1010#step:8:14)12e0f0f] Update Sphinx Documentation
4 files changed, 3 insertions(+), 3 deletions(-)
To https://github.com/RuminantFarmSystems/MASM.git
! [rejected] HEAD -> eds_trial_sphinx_updater (fetch first)
error: failed to push some refs to 'https://github.com/RuminantFarmSystems/MASM.git'
hint: Updates were rejected because the remote contains work that you do not
hint: have locally. This is usually caused by another repository pushing to
hint: the same ref. If you want to integrate the remote changes, use
hint: 'git pull' before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
Error: Invalid status code: 1
at ChildProcess. (/home/runner/work/_actions/actions-js/push/master/start.js:9:19)
at ChildProcess.emit (node:events:513:28)
at maybeClose (node:internal/child_process:1100:[16](https://github.com/RuminantFarmSystems/MASM/actions/runs/7103107082/job/19335012054?pr=1010#step:8:17))
at Process.ChildProcess._handle.onexit (node:internal/child_process:304:5) {
code: 1
}
Error: Invalid status code: 1
at ChildProcess. (/home/runner/work/_actions/actions-js/push/master/start.js:9:[19](https://github.com/RuminantFarmSystems/MASM/actions/runs/7103107082/job/19335012054?pr=1010#step:8:20))
at ChildProcess.emit (node:events:513:28)
at maybeClose (node:internal/child_process:1100:16)
at Process.ChildProcess._handle.onexit (node:internal/child_process:[30](https://github.com/RuminantFarmSystems/MASM/actions/runs/7103107082/job/19335012054?pr=1010#step:8:31)4:5)

@ew3361zh I am not seeing an error there either - it looks like it didn't run any of the Actions for some reason.
But it says in the error message on this branch 
```
Push to branch eds_trial_sphinx_updater
...
! [rejected] HEAD -> eds_trial_sphinx_updater (fetch first)
```
so somehow stuff from 1020 got in here.

**Review by Isgotkowitz — APPROVED** [review]

File: .github/workflows/sphinx.yml (hunk 1 above)
Comment (allisterakun) [reply .github/workflows/sphinx.yml]: I would agree with Ed about excluding the unit tests. I think all the unit tests I have seen so far are self-descriptive enough; we don't need explicit documentation.

**Review by allisterakun — APPROVED** [review]
Great job, Niko! All changes in the GitHub action are straight-forward.

File: .github/workflows/sphinx.yml (hunk 1 above)
Comment (ew3361zh) [reply .github/workflows/sphinx.yml]: Thank you @allisterakun and @Isgotkowitz, I will add Ed's line and make sure it's running as intended before merging.

**Pull request merged**